airbyte-integrations/connectors/source-notion/acceptance-test-config.yml

```diff
@@ -33,6 +33,7 @@ acceptance_tests:
         configured_catalog_path: "integration_tests/incremental_catalog.json"
         future_state:
           future_state_path: "integration_tests/abnormal_state.json"
+        skip_comprehensive_incremental_tests: true
```

Comment:
Glad you asked! This test is a frequent pain point for connectors with limited sandbox data and can throw failures even when nothing is wrong; at first glance I had assumed this was just another such case, but I took a deeper look into the test results and am seeing some fishy logs for the Blocks stream. Investigating further and will get back to you soon as I have a clearer sense of what's going on here.

Reply:
@cnoll Ok, I think I understand what's going on. The Blocks stream is a doozy, as we not only have no way of filtering or sorting results, but due to the nature of the recursive traversal through the blocks hierarchy, there isn't a defined top-level of records used in slicing. 

Instead, an id_stack is populated with the parent stream `ids`; then as we query those pages, the nested blocks have their `ids` added to the stack and become the next slice (which will contain child blocks that will become slices for their own children, and down the rabbit hole we go). Right now the stream tracks the `cursor_value` of each record and uses a custom `StateValueWrapper` to compare and update the `max_cursor_time` until there is only one `id` remaining in the stack; this condition flips an `is_finished` boolean, which then triggers the final calculation of `max_cursor` in the `StateValueWrapper` to actually update the stream's state.

Here's some print statements of what these values looks like when there is an existing state value:

```
current stream_state: {'last_edited_time': 2022-10-10T04:00:00.000Z}
current StateValueWrapper: stream=<source_notion.streams.Blocks object at 0x7fc71ff9a140> state_value='2022-10-10T04:00:00.000Z' max_cursor_time='2023-10-10T13:51:00.000Z' <- `max_cursor_time` represents the latest cursor value in the read thus far

most recent record_cursor: 2023-10-23T17:30:00.000Z
updated StateValueWrapper.max_cursor_time: 2023-10-23T17:30:00.000 <- this will keep updating and become the new value of state only if `self.is_finished` has been flipped
```

So like with the other Notion streams, it's functionally a full read operation that uses filtering on subsequent reads to prevent records from emitting. The issue exposed by the `comprehensive_incremental_tests` is that the value of state is only updated at the end of the read, rather than on each call to `get_updated_state`. Unfortunately, changing the logic to update the value of state after a slice is processed would be dangerous, since our client-side filtering logic would then start taking effect on the first read and exclude unsorted blocks records that have never been synced.

Ordinarily, the best solution would probably be to set up per-partition semi-incremental logic (like in the Comments stream) so we can update each slice's state as we go and aren't reliant on a first successful sync to set a value for `stream_state`. My one hesitation in this case is with the inevitable ballooning of the state object for this stream. For reference, our tiny sandbox account with just a few basic sample pages has over 500 blocks records, any number of which could be a potential parent/slice for other blocks. I'm not certain what the memory limitations are for state objects, but I'm concerned about the potential scaling implications on larger Notion accounts, as well as the effect on debugging logs. Any thoughts on how to handle this? I'm tempted to say "working as intended!" for now since this is how the connector has always worked, but I'm admittedly biased by a desire to have this ready to ship by Monday 😅

Reply:
I assume the mention should go to @clnoll

Reply:
Whoops haha yes, so sorry!

Reply:
Wow, thank you for the detailed explanation! This is super helpful.

> The issue exposed by the comprehensive_incremental_tests is that the value of state is only updated at the end of the read, rather than on each call to get_updated_state.

It sounds like the main side effect is that we may end up syncing some records more than necessary. Is that right?

In any case I agree with you about the best solution, but also that since this is how it's always worked, we can leave it as-is for now. Would you mind creating an issue for this, in case we decide we want to follow up on it?

Reply:
Yup, that sounds right: if we get halfway through a first sync and there's an interruption, any records that were successfully synced will still be emitted on the subsequent attempt. Put the issue together here: https://github.com/airbytehq/airbyte/issues/36698, just a copy/paste of the investigation to start but I'll do some grooming 👍